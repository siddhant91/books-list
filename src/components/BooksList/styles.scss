@import '../../styles/main.scss';
.bokl-books-listing {
	&--title {
		color: $bokl-color-graphite;
	}
	&--list__card {
		padding: $bokl-spacing-3;
		margin: $bokl-spacing-3 $bokl-spacing-3 $bokl-spacing-3 0;
		width: 100%;
		@media only screen and (min-width: $bokl-breakpoint-desktop) {
			width: 49%;
		}
		&--title {
			color: $bokl-color-graphite;
		}
		&--details {
			color: $bokl-color-grey-4;
			&__label {
				margin-right: $bokl-spacing-2;
			}
		}
	}
}
