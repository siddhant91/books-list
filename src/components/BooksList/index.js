import { arrayOf, shape } from 'prop-types';

// Components

import Card from '../Card';

// Styles
import './styles.scss';

const BooksList = ({ booksData }) => {
	const getCardDetails = ({ label, value }) => {
		return (
			<p className="bokl-books-listing--list__card--details mb-1">
				<span className="bokl-books-listing--list__card--details__label">{label}</span>
				<span>{value || 'N/A'}</span>
			</p>
		);
	};

	const getBookCards = () => {
		if (booksData.length) {
			const bookCards = booksData.map((book) => {
				const {
					volumeInfo: { title, authors = [], publisher = '', publishedDate = '' } = {},
				} = book;
				return (
					<Card additionalClasses="bokl-books-listing--list__card" key={title}>
						<h4 className="bokl-books-listing--list__card--title">{title}</h4>
						{getCardDetails({
							label: 'Authors:',
							value: authors.length > 0 ? authors.toString() : 'N/A',
						})}
						{getCardDetails({ label: 'Publisher:', value: publisher })}
						{getCardDetails({ label: 'Publishe Date:', value: publishedDate })}
					</Card>
				);
			});
			return bookCards;
		}
		return <p>Sorry no books to Display</p>;
	};

	return (
		<div className="bokl-books-listing">
			<h2 className="bokl-books-listing--title">All Books</h2>
			<div className="bokl-books-listing--list mt-5 d-flex flex-wrap justify-content-between">
				{getBookCards()}
			</div>
		</div>
	);
};

BooksList.propTypes = {
	booksData: arrayOf(shape({})),
};

BooksList.defaultProps = {
	booksData: [],
};

export default BooksList;
