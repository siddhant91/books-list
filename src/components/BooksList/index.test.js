import { render, screen, cleanup } from '@testing-library/react';

import BooksList from './index';

afterEach(cleanup);

const renderComponent = (booksData = []) => {
	return <BooksList booksData={booksData} />;
};

it('Matches snapshot', () => {
	const { asFragment } = render(renderComponent());
	expect(asFragment()).toMatchSnapshot();
});

it('renders efault title', async () => {
	render(renderComponent());
	expect(await screen.findByText('All Books')).toBeInTheDocument();
});

it('renders no book to display text', async () => {
	render(renderComponent());
	expect(await screen.findByText('Sorry no books to Display')).toBeInTheDocument();
});

it('renders books list', async () => {
	const mockData = [
		{
			volumeInfo: {
				title: 'XYZ',
				authors: ['A', 'B'],
				publisher: 'ABC',
				publishedDate: '2021-01-16',
			},
		},
	];
	render(renderComponent(mockData));
	const bookCards = await screen.findAllByTestId('bokl-card');
	expect(bookCards).toHaveLength(1);
});
