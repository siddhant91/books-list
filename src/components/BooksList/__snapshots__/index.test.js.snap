// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bokl-books-listing"
  >
    <h2
      class="bokl-books-listing--title"
    >
      All Books
    </h2>
    <div
      class="bokl-books-listing--list mt-5 d-flex flex-wrap justify-content-between"
    >
      <p>
        Sorry no books to Display
      </p>
    </div>
  </div>
</DocumentFragment>
`;
